packages/fork-choice/src/forkChoice/forkChoice.ts

```diff
@@ -214,10 +214,17 @@ export class ForkChoice implements IForkChoice {
           root: fromHexString(headRoot),
         });
       }
+      const blockSummary = toBlockSummary(headNode);
 
-      return (this.head = toBlockSummary(headNode));
+      const isSameHead =
+        blockSummary && this.head && toHexString(blockSummary.blockRoot) == toHexString(this.head.blockRoot);
+      const isSameChain =
+        isSameHead || (this.head && blockSummary && this.isDescendant(this.head.blockRoot, blockSummary.blockRoot));
+      Object.assign(mLabels, {resStatus: isSameHead ? "samehead" : isSameChain ? "newhead" : "newfork"});
+
+      return (this.head = blockSummary);
     } finally {
-      if (timer) timer();
+      if (timer) timer(mLabels);
```

Comment:
This metrics are way too mixed up. You should use different metrics to answer different questions:
- How often is updateHead() called?
- How much cpu time does the fork choice routine require?
- How many errors happen on the fork choice routine?

Also, could you describe what type of analysis or question are you trying to answer with the `isSameHead` and `isSameChain` tags? We must not do any extra work if metrics are disabled and running `toHexString` is not cheap.

Reply:
> 
> 
> This metrics are way too mixed up. You should use different metrics to answer different questions:
> 
>     * How often is updateHead() called?
> 
>     * How much cpu time does the fork choice routine require?
> 
>     * How many errors happen on the fork choice routine?
> 
> 
> Also, could you describe what type of analysis or question are you trying to answer with the `isSameHead` and `isSameChain` tags? We must not do any extra work if metrics are disabled and running `toHexString` is not cheap.

yea, isSameHead and isSameChain should go inside a check, will fix that part. 

However, my opinion on metrics is, use the prom features to maximum extend as this is just one collection call solving 4 metrics, instead of having 4 separate collection calls. Our discussion in the call was also on the same lines. I would want prom to do more work than have more code/calls here. Infact for prom, it would be just a matter of building index on the fly on labels, a little costlier for prom for sure. 

Do let me know your thoughts on this. @dapplion  @wemeetagain

Reply:
> 
> Also, could you describe what type of analysis or question are you trying to answer with the `isSameHead` and `isSameChain` tags? 

beacon_fork_choice_changed_head_total -- via  beacon_fork_choice_find_head_seconds_count{resStatus="newhead"}
beacon_fork_choice_reorg_total-- via  beacon_fork_choice_find_head_seconds_count{resStatus="newfork"}

kindly check description for other mappings.

Reply:
> However, my opinion on metrics is, use the prom features (labels) to maximum extent as this is just one collection call solving 4 metrics, instead of having 4 separate collection calls. 

Having two different metrics `metric_a` and `metric_b` is equivalent to having one metric with two labels `metric{x="a"}` and `metric{x="b"}`.

We should only use labels where it makes sense to plot data together, on when the label values are not known until runtime.

Reply:
I don't understand why would you want to count errors with a histogram metric

Reply:
ummm actually I don't have a strong opinion about not using histogram and in favor of using gauge (separate metric for counts).

However I do think labels in counting for {syncStatus,resStatus} makes total sense. Labels are dimensions not a metric  or in non data science terminology they are filters. I don't agree with an assessment that label values have to be run time dynamic values. I believe that Prometheus is made to natively handle the dimensions. Plus One collect/observe/timer call is better than 3. 

May be can we have a call on this so as to discuss and make an informed decision? on monday at your convenience or same time when we have update calls?

Reply:
@g11in If you look at the original issue `beacon_fork_choice_errors_total` is a gauge. Take a look at other eth2 clients approaches and would see no-one uses the approach you propose

Reply:
> @g11in If you look at the original issue `beacon_fork_choice_errors_total` is a gauge. Take a look at other eth2 clients approaches and would see no-one uses the approach you propose

may be i wasn't clear in my previous comment: 
i donnot mind having separate metric gauge for tracking counts, what I am saying is the use of labels on that gauge metric to give us syncStatus, resStatus combinations.. thats what i want discussion on.

Reply:
I understand but I would highly suggest to stick to the original proposal https://github.com/ChainSafe/lodestar/issues/2558 and
- use a histogram to track time duration
- use a gauge to count events

Reply:
sure, no problem. will make the changes and update PR 👍